File: Domains/Apps/Workspace/Typescript/Angular/src/allors/domain/base/database/query/Like.ts
import { RoleType } from '../../../../meta';
import { Predicate } from './Predicate';

export class Like implements Predicate {
  roleType: RoleType;
  value: any;

  constructor(fields?: Partial<Like>) {
    Object.assign(this, fields);
  }

  toJSON(): any {
    return {
      _T: 'Like',
      rt: this.roleType.id,
      v: this.value,
    };
  }
}

File: Core/Workspace/CSharp/Domain/Export/Core/Data/Like.cs
//-------------------------------------------------------------------------------------------------
// <copyright file="Like.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>
//-------------------------------------------------------------------------------------------------

namespace Allors.Workspace.Data
{
    using Allors.Workspace.Meta;
    using System.Collections.Generic;
    using Allors.Protocol.Data;

    public class Like : IRolePredicate
    {
        public Like(IRoleType roleType = null) => this.RoleType = roleType;

        public IRoleType RoleType { get; set; }

        public string Value { get; set; }

        public string Parameter { get; set; }

        public Predicate ToJson() =>
            new Predicate
            {
                Kind = PredicateKind.Like,
                RoleType = this.RoleType?.Id,
                Value = DataConvert.ToString(this.Value),
                Parameter = this.Parameter,
            };

        bool IPredicate.ShouldTreeShake(IReadOnlyDictionary<string, object> arguments) => ((IPredicate)this).HasMissingArguments(arguments);

        bool IPredicate.HasMissingArguments(IReadOnlyDictionary<string, object> arguments) => this.Parameter != null && (arguments == null || !arguments.ContainsKey(this.Parameter));
    }
}
